buidl/message.py

```diff
@@ -0,0 +1,224 @@
+        elif(to_sign.tx_outs[0].script_pubkey.commands != [106]):
+            raise ValueError("ScriptPubKey incorrect", to_sign.tx_outs[0].script_pubkey)
+        else:
+            return to_sign
+            
+    else:
+        # signature is either None or an encoded witness stack
+        
+        # Identifies the index of the output from the virtual to_spend tx to be "spent"
+        prevout_index = 0
+        sequence = 0
+        # TxInput identifies the single output from the to_spend Tx
+        tx_input = TxIn(to_spend_tx_hash,prevout_index,script_sig=None,sequence=sequence)
+    
+        value = 0
+        # OP Code 106 for OP_RETURN
+        commands = [106]
+        scriptPubKey = Script(commands)
+
+        tx_output = TxOut(value,scriptPubKey)
+        locktime=0
+        version=0
+        tx_inputs = [tx_input]
+        tx_outputs = [tx_output]
+        network="mainnet"
+        # is to_sign always a segwit?
+        segwit=True
+    
+        # create unsigned to_sign transaction
+        to_sign_tx = Tx(version,tx_inputs,tx_outputs,locktime,network,segwit)
+        
+        if sig_bytes:
+            try:
+                stream = io.BytesIO(sig_bytes)
+                witness = Witness.parse(stream)
+                # Set the witness on the to_sign tx input
+                to_sign_tx.tx_ins[0].witness = witness
+            except:
+                # TODO: Fall back to legacy ...
+                print("Signature is niether an encoded witness or full transaction. Fall back to legacy")
+                return None
+                
+        return to_sign_tx
+
+# Test if sig_bytes can be decoded to a transaction
+# TODO: Is there a better way to test than this?
+def is_full_signature(sig_bytes):
+    try:
+        sig_stream = io.BytesIO(sig_bytes)
+
+        Tx.parse(sig_stream)
+    # TODO: more specific exception handling
+    except:
+        return False
+    return True
+        
+
+def sign_message(format: MessageSignatureFormat, private_key: PrivateKey, address: str, message: str):
+    
+    if (format != MessageSignatureFormat.LEGACY):
+        return sign_message_bip322(format,private_key,address,message)
+
+    script_pubkey = address_to_script_pubkey(address)
+    
+    if (not script_pubkey.is_p2pkh()):
+        raise ValueError("Address must be p2pkh for LEGACY signatures")
+        
+    # TODO: This legacy signing needs to produce a compact encoding of signature AND public key
+    # Needs implementing in the library I believe?
+    raise NotImplementedError("Legacy signing not yet implemented. Require compact encoding of sig and pubkey")
+    # b_msg = str_to_bytes(message)
+    # signature = private_key.sign_message(b_msg)
+    
+    # return base64_encode(signature.der())    
+    
+def sign_message_bip322(format: MessageSignatureFormat, private_key: PrivateKey, address: str, message: str):
+    
+    assert(format != MessageSignatureFormat.LEGACY)
```

Comment:
A helpful error message would be nice.

Reply:
Added the following error message to the assert

`BIP 322 Signatures must be represented in either the Simple of Full Formats`